### PR TITLE
[draft]curvefs/client: opt of bs for fs storage backend

### PR DESCRIPTION
**Pull request by wuhongsong**
### What problem does this PR solve?

Issue Number: #1426  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by YunhuiChen** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by YunhuiChen** [comment]
cicheck

**Comment by YunhuiChen** [comment]
cicheck

**Comment by wuhongsong** [comment]
i have a small cluster just have 3 disks. and the bandwith is 50MB/s，follow  write: IOPS=103, BW=51.7MiB/s (54.2MB/s)(19.0GiB/395868msec);

and after opt,we can get double bandwith，follow write: IOPS=192, BW=96.1MiB/s (101MB/s)(20.0GiB/213048msec); 0 zone resets , and the disks util is nearly 100%

**Comment by ilixiaocui** [comment]
@wuhongsong Please deal with the confilict.

File: curvefs/src/client/volume/client_volume_adaptor.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        VLOG(0) << "volume flush dataCache err, inode: "
+            << inodeId<< ", offset: " << inodeOffset
+            << ", length: " << len;
+        return ret;
+    }
+
+    VLOG(9) << "volume flush dataCache end, inode: "
+       << inodeId<< ", offset: " << inodeOffset
+       << ", length: " << len;
+// 这个正确么？
```
Comment (aspirer) [inline curvefs/src/client/volume/client_volume_adaptor.cpp:145]: ?

File: curvefs/src/client/volume/fuse_volume_client.cpp [hunk 2]
```diff
@@ -83,151 +77,136 @@ CURVEFS_ERROR FuseVolumeClient::FuseOpInit(void *userdata,
...
-        fsMetric_->userWrite.qps.count << 1;
-        fsMetric_->userWrite.latency << timer.u_elapsed();
-        fsMetric_->userWriteIoSize.set_value(size);
-    }
+    inodeManager_->ShipToFlush(inodeWrapper);
 
-    VLOG(9) << "write end, ino: " << ino << ", offset: " << off
-            << ", length: " << size << ", written: " << *wSize;
+    if (fi->flags & O_DIRECT || fi->flags & O_SYNC || fi->flags & O_DSYNC) {
+        // Todo: do some cache flush later
```
Comment (aspirer) [inline curvefs/src/client/volume/fuse_volume_client.cpp:138]: add some logs for this situation?

File: curvefs/src/client/volume/fuse_volume_client.cpp [hunk 3]
```diff
@@ -83,151 +77,136 @@ CURVEFS_ERROR FuseVolumeClient::FuseOpInit(void *userdata,
...
+CURVEFS_ERROR FuseVolumeClient::FuseOpWrite(fuse_req_t req, fuse_ino_t ino,
+        const char *buf, size_t size, off_t off,
+        struct fuse_file_info *fi,
+        size_t *wSize) {
+    // check align
     if (fi->flags & O_DIRECT) {
         if (!(is_aligned(off, DirectIOAlignment) &&
-              is_aligned(size, DirectIOAlignment))) {
-            fsMetric_->userWrite.eps.count << 1;
+              is_aligned(size, DirectIOAlignment)))
```
Comment (aspirer) [inline curvefs/src/client/volume/fuse_volume_client.cpp:97]: why delete userWrite.eps.count?

File: curvefs/src/client/volume/fuse_volume_client.cpp [hunk 4]
```diff
@@ -83,151 +77,136 @@ CURVEFS_ERROR FuseVolumeClient::FuseOpInit(void *userdata,
...
+    CURVEFS_ERROR ret = inodeManager_->GetInode(ino, inodeWrapper);
     if (ret != CURVEFS_ERROR::OK) {
-        if (fsMetric_) {
-            fsMetric_->userWrite.eps.count << 1;
-        }
-        LOG(ERROR) << "write error, ino: " << ino << ", offset: " << off
-                   << ", len: " << size
-                   << ", error: " << ret;
+        LOG(ERROR) << "inodeManager get inode fail, ret = " << ret
+                   << ", inodeid = " << ino;
```
Comment (aspirer) [inline curvefs/src/client/volume/fuse_volume_client.cpp:119]: ditto

File: curvefs/src/client/volume/fuse_volume_client.cpp [hunk 5]
```diff
@@ -83,151 +77,136 @@ CURVEFS_ERROR FuseVolumeClient::FuseOpInit(void *userdata,
...
+    inodeManager_->ShipToFlush(inodeWrapper);
 
-    VLOG(9) << "write end, ino: " << ino << ", offset: " << off
-            << ", length: " << size << ", written: " << *wSize;
+    if (fi->flags & O_DIRECT || fi->flags & O_SYNC || fi->flags & O_DSYNC) {
+        // Todo: do some cache flush later
+    }
 
-    return CURVEFS_ERROR::OK;
+    if (enableSumInDir_ && changeSize != 0) {
```
Comment (aspirer) [inline curvefs/src/client/volume/fuse_volume_client.cpp:141]: are these codes copied from s3 client?

File: curvefs/src/client/volume/fuse_volume_client.cpp [hunk 6]
```diff
@@ -83,151 +77,136 @@ CURVEFS_ERROR FuseVolumeClient::FuseOpInit(void *userdata,
...
+                                       size_t size, off_t off,
+                                       struct fuse_file_info *fi, char *buffer,
+                                       size_t *rSize) {
     // check align
     if (fi->flags & O_DIRECT) {
         if (!(is_aligned(off, DirectIOAlignment) &&
-              is_aligned(size, DirectIOAlignment))) {
-            fsMetric_->userRead.eps.count << 1;
-
+              is_aligned(size, DirectIOAlignment)))
```
Comment (aspirer) [inline curvefs/src/client/volume/fuse_volume_client.cpp:165]: ditto

File: curvefs/src/client/volume/fuse_volume_client.cpp [hunk 7]
```diff
@@ -83,151 +77,136 @@ CURVEFS_ERROR FuseVolumeClient::FuseOpInit(void *userdata,
...
+    uint64_t fileSize = inodeWrapper->GetLength();
 
-    if (fsMetric_) {
-        fsMetric_->userRead.bps.count << size;
-        fsMetric_->userRead.qps.count << 1;
-        fsMetric_->userRead.latency << timer.u_elapsed();
-        fsMetric_->userReadIoSize.set_value(size);
+    size_t len = 0;
+    if (fileSize <= off) {
+        *rSize = 0;
```
Comment (aspirer) [inline curvefs/src/client/volume/fuse_volume_client.cpp:181]: eps count

File: curvefs/src/client/client_storage_adaptor.cpp [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+            waitInterval_.StopWait();
+            fsCacheManager_->WaitFlush();
+        }
+    }
+    uint64_t memCacheRatio = fsCacheManager_->MemCacheRatio();
+    int64_t exceedRatio = memCacheRatio - memCacheNearfullRatio_;
+    if (exceedRatio > 0) {
+        // offer to do flush
+        waitInterval_.StopWait();
+        // upload to s3 derectly or cache disk full
```
Comment (aspirer) [inline curvefs/src/client/client_storage_adaptor.cpp:148]: change comments. not only s3 now.

File: curvefs/src/client/client_storage_adaptor.cpp [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+        if ((size + pendingReq * fuseMaxSize_) >= maxSize) {
+            VLOG(6) << "write cache is full, wait flush. size: " << size
+                    << ", maxSize:" << maxSize;
+            // offer to do flush
+            waitInterval_.StopWait();
+            fsCacheManager_->WaitFlush();
+        }
+    }
+    uint64_t memCacheRatio = fsCacheManager_->MemCacheRatio();
+    int64_t exceedRatio = memCacheRatio - memCacheNearfullRatio_;
```
Comment (aspirer) [inline curvefs/src/client/client_storage_adaptor.cpp:144]: here is memcache, but there is diskcache blow.

File: curvefs/src/client/client_storage_adaptor.cpp [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+    if (exceedRatio > 0) {
+        // offer to do flush
+        waitInterval_.StopWait();
+        // upload to s3 derectly or cache disk full
+        bool needSleep =
+            (DisableDiskCache() || IsReadCache()) ||
+            (IsReadWriteCache() && diskCacheManagerImpl_->IsDiskCacheFull());
```
Comment (aspirer) [inline curvefs/src/client/client_storage_adaptor.cpp:151]: diskcache or memcache?

**Pull request closed without merging**